### PR TITLE
Messaging rewrite

### PR DESCRIPTION
**Pull request by mmermerkaya**
I didn't want to leave the FCM SDK in it's current state, so I rewrote a large portion of it (and hopefully without any API changes this time :smile:). There are also some API changes that I'd really like to do, like removing `getToken` and `deleteToken` from `SWController` or making `vapidKey` a parameter of `getToken`, but I guess they'll have to wait.

**Changes:**
- Composition over inheritance
  - Removed `BaseController` that the two main controllers were extending and moved the common parts to "token manager" module.
  - Removed `DbInterface`, which was not actually an interface but an abstract class, and the two classes that were extending it.
- Database
  - Moved Push API related fields to the `subscriptionOptions` subfield. This is now an optional field, as preparation for the eventual Safari support, which will not use the Push API.
  - Dropped `swScope` as an index, for the same reason as before.
  - Database key changed to App ID, as there should only be one token for every App ID. This should prevent "multiple messages received" issues in topic sends. 
  - Created a new DB and added a migration function because the object store needed to change. This allowed removing the old DB code.
  - Started using the battle-tested IndexedDB library [idb](https://github.com/jakearchibald/idb/), which is already being used in Installations.
- Size reduction
  - Non minified ES5 ESM build is now 70.8kB. Old one was 96.8kB.
  - Minified and gzipped ES5 build is now 10.1kB. Old one was 12.5kB.
- Tests
  - Increased coverage from 93% to 99%. Coveralls reports a drop in coverage because I also changed the coverage config to stop including test files in coverage reports.
  - Made the tests more meaningful by removing all private member access from tests. Focusing on testing behavior, not implementation.
  - Added complete Service Worker fakes to use with every test. Removed messing with prototype objects in fakes.
  - Removed the uses of `any` type from tests. There used to be 83 instances of type `any` in tests.
- Structure and Naming:
  - Greatly reduced the number of classes. Not everything needs to be a class.
  - Combined src/ and test/ folders, following our style guide.
  - Removed the -model suffix. More meaningful names overall.

### CODE REVIEW

File: config/webpack.test.js [hunk 1]
```diff
@@ -50,7 +50,7 @@ module.exports = {
           options: { esModules: true }
         },
         enforce: 'post',
-        include: path.resolve(process.cwd(), 'src')
+        exclude: [/\.test\.ts$/, /test(ing)?\//]
```
Comment (dwoffinden) [inline config/webpack.test.js]: presumably this would also exclude a directory called "nottesting" :stuck_out_tongue: 

would a \b "word boundary" work?

/\btest(ing)?\//

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Assertions

File: packages/messaging/src/controllers/sw-controller.ts [hunk 2]
```diff
@@ -1,6 +1,6 @@
 /**
  * @license
- * Copyright 2017 Google Inc.
+ * Copyright 2019 Google Inc.
```
Comment (dwoffinden) [inline packages/messaging/src/controllers/sw-controller.ts]: go/copyright

It seems like this shouldn't be replaced, and you can add the current year so long as you preserve the original year? So 2017 or 2017-2019 but not 2019?

File: packages/messaging/src/core/api.test.ts [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+        new Response(JSON.stringify({ error: { message: 'error message' } }))
+      );
+      await expect(
+        requestUpdateToken(firebaseDependencies, tokenDetails)
+      ).to.be.rejectedWith('messaging/token-update-failed');
+
+      fetchStub.resolves(
+        new Response(
+          JSON.stringify({
+            /* no token */
```
Comment (dwoffinden) [inline packages/messaging/src/core/api.test.ts]: You could replace these four lines with the four characters "{}" but I guess consistency is important too :stuck_out_tongue:

File: packages/messaging/src/controllers/window-controller.ts [hunk 4]
```diff
@@ -1,6 +1,6 @@
 /**
  * @license
- * Copyright 2017 Google Inc.
+ * Copyright 2019 Google Inc.
```
Comment (dwoffinden) [inline packages/messaging/src/controllers/window-controller.ts]: Again with the copyright

I'll stop mentioning it but there are other files too :stuck_out_tongue:

File: packages/messaging/src/controllers/window-controller.ts [hunk 5]
```diff
@@ -86,254 +92,144 @@ export class WindowController extends BaseController {
...
+      this.onMessageCallback(payload);
+    }
+
+    const { data } = payload;
+    if (
+      data &&
+      FN_CAMPAIGN_ID in data &&
+      data[FN_CAMPAIGN_ANALYTICS_ENABLED] === '1'
+    ) {
+      // This message has a campaign id, meaning it was sent using the FN Console.
```
Comment (dwoffinden) [inline packages/messaging/src/controllers/window-controller.ts]: The branding seems to be "Firebase Cloud Messaging" / "the Firebase Console" / "the Notifications composer"

https://firebase.google.com/docs/cloud-messaging/js/send-with-console

So if you want the comment to make sense to someone who doesn't know what "FN" means, I'd just say "the Firebase Console" or "the Notifications composer".

I guess the FN_... constants should just be COMPOSER or CONSOLE?

Also "id" -> "ID" (we aren't talking about https://en.wikipedia.org/wiki/Id,_ego_and_super-ego#Id)

File: packages/messaging/src/core/api.ts [hunk 6]
```diff
@@ -0,0 +1,186 @@
+/**
+ * @license
+ * Copyright 2017 Google Inc.
```
Comment (dwoffinden) [inline packages/messaging/src/core/api.ts]: ok if this is a new file then it *should* use 2019 :smile:

File: packages/messaging/src/core/token-management.ts [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+import { dbGet, dbSet, dbRemove } from '../helpers/idb-manager';
+import { FirebaseInternalDependencies } from '../interfaces/internal-dependencies';
+import { SubscriptionOptions } from '../interfaces/subscription-options';
+import { TokenDetails } from '../interfaces/token-details';
+import { requestUpdateToken, requestGetToken, requestDeleteToken } from './api';
+import { arrayToBase64 } from '../helpers/array-to-base64';
+import { ERROR_FACTORY, ErrorCode } from '../util/errors';
+
+/** UpdateRegistration will be called once every week. */
+const TOKEN_EXPIRATION_MS = 7 * 24 * 60 * 60 * 1000; // 7 days
```
Comment (dwoffinden) [inline packages/messaging/src/core/token-management.ts]: So much sadness from the lack of a ubiquitous date/time api :(

File: packages/messaging/src/helpers/array-to-base64.test.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+import { arrayToBase64 } from './array-to-base64';
+import { expect } from 'chai';
+import '../testing/setup';
+
+describe('arrayToBase64', () => {
+  it('returns expected value for p256dh example', () => {
+    // prettier-ignore
+    const buffer = new Uint8Array([
+      4, 181, 98, 240, 48, 62, 75, 119, 193, 227, 154, 69, 250, 216, 53, 110,
```
Comment (dwoffinden) [inline packages/messaging/src/helpers/array-to-base64.test.ts]: Curious: any particular reason for these numbers? A valid test key? Lottery numbers?

For test code, if the exact values don't matter, it's usually an excuse to use joke data that's obviously unimportant, or at least exercises the edge cases :p

File: packages/messaging/src/helpers/idb-manager.ts [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+// Exported for tests.
+export const DATABASE_NAME = 'firebase-messaging-database';
+const DATABASE_VERSION = 1;
+const OBJECT_STORE_NAME = 'firebase-messaging-store';
+
+let dbPromise: Promise<DB> | null = null;
+function getDbPromise(): Promise<DB> {
+  if (!dbPromise) {
+    dbPromise = openDb(DATABASE_NAME, DATABASE_VERSION, upgradeDB => {
+      // We don't use 'break' in this switch statement, the fall-through
```
Comment (dwoffinden) [inline packages/messaging/src/helpers/idb-manager.ts]: Solid logic, if a little confusing when you only have one case :smile:

File: packages/messaging/src/helpers/idb-manager.ts [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+
+// Exported for tests.
+export const DATABASE_NAME = 'firebase-messaging-database';
+const DATABASE_VERSION = 1;
+const OBJECT_STORE_NAME = 'firebase-messaging-store';
+
+let dbPromise: Promise<DB> | null = null;
+function getDbPromise(): Promise<DB> {
+  if (!dbPromise) {
+    dbPromise = openDb(DATABASE_NAME, DATABASE_VERSION, upgradeDB => {
```
Comment (dwoffinden) [inline packages/messaging/src/helpers/idb-manager.ts]: upgradeDb

(else, ought the method not be getDBPromise?)

File: packages/messaging/src/helpers/migrate-old-database.ts [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+  createTime: number;
+}
+
+const OLD_DB_NAME = 'fcm_token_details_db';
+/**
+ * The last DB version of 'fcm_token_details_db' was 4. This is one higher,
+ * so that the upgrade callback is called for all versions of the old DB.
+ */
+const OLD_DB_VERSION = 5;
+const OLD_OBJECT_STORE_NAME = 'fcm_token_object_Store';
```
Comment (dwoffinden) [inline packages/messaging/src/helpers/migrate-old-database.ts]: Why capital "S"?

File: packages/messaging/src/interfaces/message-payload.ts [hunk 12]
```diff
@@ -29,5 +27,5 @@ export interface FcmOptions {
 export interface MessagePayload {
   fcmOptions?: FcmOptions;
   notification?: NotificationDetails;
-  data?: { [key: string]: unknown };
+  data?: any; // eslint-disable-line @typescript-eslint/no-explicit-any
```
Comment (dwoffinden) [inline packages/messaging/src/interfaces/message-payload.ts]: Isn't object better than any, in that it forces you to check whether anything is there rather than silently becoming anything?

File: packages/messaging/src/testing/compare-headers.ts [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+// so we need to do that manually.
+export function compareHeaders(
+  expectedHeaders: HeadersWithEntries,
+  actualHeaders: HeadersWithEntries
+): void {
+  expect(expectedHeaders.entries).not.to.be.undefined;
+  expect(actualHeaders.entries).not.to.be.undefined;
+
+  const expected = new Map(Array.from(expectedHeaders.entries!()));
+  const actual = new Map(Array.from(actualHeaders.entries!()));
```
Comment (dwoffinden) [inline packages/messaging/src/testing/compare-headers.ts]: ISTM like a little helper to assert that the entries are there and then turn them into a Map would be useful, you could then call it for both sets of headers :p

File: packages/messaging/src/testing/compare-headers.ts [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+// Chai doesn't check if Headers objects contain the same entries,
+// so we need to do that manually.
+export function compareHeaders(
+  expectedHeaders: HeadersWithEntries,
+  actualHeaders: HeadersWithEntries
+): void {
+  expect(expectedHeaders.entries).not.to.be.undefined;
+  expect(actualHeaders.entries).not.to.be.undefined;
+
+  const expected = new Map(Array.from(expectedHeaders.entries!()));
```
Comment (dwoffinden) [inline packages/messaging/src/testing/compare-headers.ts]: Doesn't the Map constructor take an iterable why do you need Array.from()?

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map

File: packages/messaging/src/testing/fakes/firebase-dependencies.ts [hunk 15]
```diff
@@ -47,24 +53,26 @@ export function makeFakeApp(options: FirebaseOptions = {}): FirebaseApp {
...
-export function makeFakeAnalyticsProvider(): Provider<
-  FirebaseAnalyticsInternalName
-> {
-  return new Provider<FirebaseAnalyticsInternalName>(
-    'analytics-internal',
-    new ComponentContainer('test')
-  );
+function getFakeAnalyticsProvider(): Provider<FirebaseAnalyticsInternalName> {
+  const analytics: FirebaseAnalyticsInternal = {
+    logEvent: () => {}
```
Comment (dwoffinden) [inline packages/messaging/src/testing/fakes/firebase-dependencies.ts]: drop ":" and "=>", just "logEvent() {}"? :)

File: packages/messaging/src/util/constants.ts [hunk 16]
```diff
@@ -15,81 +15,17 @@
...
-};
+export const DEFAULT_VAPID_KEY =
+  'BDOU99-h67HcA6JeFXHbSNMu7e2yNNu3RzoMj8TM4W88jITfq7ZmPvIM1Iv-4_l2LxQcYwhqby2xGpWwzjfAnG4';
 
 export const ENDPOINT = 'https://fcmregistrations.googleapis.com/v1';
 
+/** Key of FCM Payload in Notification's data field. */
+export const FCM_MSG = 'FCM_MSG';
+
 export const FN_CAMPAIGN_ID = 'google.c.a.c_id';
```
Comment (dwoffinden) [inline packages/messaging/src/util/constants.ts]: Yeah... while we're here should these be CONSOLE or COMPOSER rather than FN?

File: config/webpack.test.js (hunk 1 above)
Comment (mmermerkaya) [reply config/webpack.test.js]: > "nottesting"

Nice footgun. :smile: 

Thanks though, I didn't know `\b`. Changed.

File: packages/messaging/src/controllers/sw-controller.ts (hunk 2 above)
Comment (mmermerkaya) [reply packages/messaging/src/controllers/sw-controller.ts]: It's a bot that adds these, I think I removed the original and it added a new one with the current (well, it was current then) year. :man_shrugging:

I'll fix them.

File: packages/messaging/src/core/api.test.ts (hunk 3 above)
Comment (mmermerkaya) [reply packages/messaging/src/core/api.test.ts]: Yeah that's true, but I think this way better expresses what I am actually doing here. Plus consistency, of course. :sweat_smile:

File: packages/messaging/src/helpers/array-to-base64.test.ts (hunk 8 above)
Comment (mmermerkaya) [reply packages/messaging/src/helpers/array-to-base64.test.ts]: No idea, it was before my time. :smile:

https://github.com/firebase/firebase-js-sdk/blob/73a586c92afe3f39a844b2be86086fddb6877bb7/packages/messaging/test/array-buffer-to-base64.test.ts

File: packages/messaging/src/helpers/idb-manager.ts (hunk 9 above)
Comment (mmermerkaya) [reply packages/messaging/src/helpers/idb-manager.ts]: I think it's still helpful to whoever's going to maintain this, telling them that in case they want to add a mutation, this is a good way to do so.

And definitely much better than having this comment, and actual break statements right afterwards. :sweat_smile:

https://github.com/firebase/firebase-js-sdk/blob/7250d25145f27ee099dc717a0130c2f8dc9841d2/packages/messaging/src/models/token-details-model.ts#L101-L102

File: packages/messaging/src/helpers/idb-manager.ts (hunk 10 above)
Comment (mmermerkaya) [reply packages/messaging/src/helpers/idb-manager.ts]: Yup, changed.

File: packages/messaging/src/helpers/migrate-old-database.ts (hunk 11 above)
Comment (mmermerkaya) [reply packages/messaging/src/helpers/migrate-old-database.ts]: No idea. It has always been like that and it always bothered me.

https://github.com/firebase/firebase-js-sdk/blob/73a586c92afe3f39a844b2be86086fddb6877bb7/packages/messaging/src/models/token-details-model.ts#L22

File: packages/messaging/src/interfaces/message-payload.ts (hunk 12 above)
Comment (mmermerkaya) [reply packages/messaging/src/interfaces/message-payload.ts]: Good point. Made it `unknown`.

File: packages/messaging/src/testing/compare-headers.ts (hunk 13 above)
Comment (mmermerkaya) [reply packages/messaging/src/testing/compare-headers.ts]: Nice.

File: packages/messaging/src/testing/compare-headers.ts (hunk 14 above)
Comment (mmermerkaya) [reply packages/messaging/src/testing/compare-headers.ts]: It worked, so yes it does. :smile:

File: packages/messaging/src/testing/fakes/firebase-dependencies.ts (hunk 15 above)
Comment (mmermerkaya) [reply packages/messaging/src/testing/fakes/firebase-dependencies.ts]: Sure.

File: packages/messaging/src/controllers/window-controller.ts (hunk 5 above)
Comment (mmermerkaya) [reply packages/messaging/src/controllers/window-controller.ts]: "Firebase Console" sounds good. Changed the constants as well.

File: packages/messaging/src/util/constants.ts (hunk 16 above)
Comment (mmermerkaya) [reply packages/messaging/src/util/constants.ts]: Changed to `CONSOLE`.

File: packages/messaging-types/index.d.ts [hunk 17]
```diff
@@ -25,7 +25,10 @@ import {
 } from '@firebase/util';
 
 export class FirebaseMessaging {
-  private constructor();
+  private constructor(app: FirebaseApp);
+  readonly app: FirebaseApp;
```
Comment (hsubox76) [inline packages/messaging-types/index.d.ts]: I don't know Typescript or this codebase that well but I noticed both classes that implement this have a constructor that takes a `FirebaseInternalDependencies` and don't have a top level `app` property, is this type generalized for another use case I'm missing?

Comment (mmermerkaya) [reply packages/messaging-types/index.d.ts]: Thanks for taking a look!

My understanding was that all Firebase services need to have an `app` field (because [this test](https://github.com/firebase/firebase-js-sdk/blob/7250d25145f27ee099dc717a0130c2f8dc9841d2/integration/shared/namespaceDefinition.json#L212) exists), so I added it here, not realizing that it is already defined in the `FirebaseService` interface [here](https://github.com/firebase/firebase-js-sdk/blob/7250d25145f27ee099dc717a0130c2f8dc9841d2/packages/app-types/private.d.ts#L38).

I changed the two classes to implement that interface instead, and removed the `app` field from this one. Please let me know if `app` (and also `INTERNAL`) isn't actually required in the core packages (there's no type checking for it AFAICT), in which case I'll remove it, since it's not actually part of the FCM API.

And I have no idea what happened to that constructor, that's definitely a mistake. :sweat_smile:

Comment (hsubox76) [reply packages/messaging-types/index.d.ts]: As far as I can tell INTERNAL and app aren't required in the code anywhere, there is one place that `service.INTERNAL.delete()` is called on services, but it checks to see that `INTERNAL` exists first. I'm not sure if we should remove the `app` getter since it seems like a staple that maybe is useful for debugging or some future plan I don't know about, and it doesn't cost much.

Comment (mmermerkaya) [reply packages/messaging-types/index.d.ts]: Alright, thanks for checking, I'll keep it then. :slightly_smiling_face:
